File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void func(int left, int right, vector<string>&ans, string res){
        if(left==0 && right == 0){
            ans.push_back(res);
        }
        if(left>0){
            func(left-1, right, ans, res+'(');
        }
        if(right>0 && left<right){
            func(left, right-1, ans, res+')');
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string res = "";
        func(n, n, ans, res);
        return ans;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    private void helper(List<String>ans, StringBuilder sb, int open, int close, int n){
        if(sb.length() == 2*n){
            ans.add(new StringBuilder(sb).toString());
            return;
        }
        // if(open>n || close>n) return;
        if(open<n){
            sb.append('(');
            helper(ans, sb, open+1, close, n);
            sb.deleteCharAt(sb.length()-1);
        }
        if(close<open){
            sb.append(')');
            helper(ans, sb, open, close+1, n);
            sb.deleteCharAt(sb.length()-1);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String>ans = new ArrayList();
        helper(ans, new StringBuilder(), 0, 0, n);
        return ans;
    }
}